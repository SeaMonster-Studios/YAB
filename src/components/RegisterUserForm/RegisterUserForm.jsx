import React from 'react'
import axios from 'axios'
import { connect } from 'react-redux'
import { bindActionCreators } from 'redux'
import PropTypes from 'prop-types'
import * as EmailValidator from 'email-validator'
//
import { setAuthCookie } from '../../utils/auth/auth'
import { testPasswordStrength } from '../../utils/input/input'
import * as authActionCreators from '../../store/actions/auth'

export class RegisterUserForm extends React.Component {
  static propTypes = {
    setAuthenticatedUser: PropTypes.func.isRequired,
  }
  state = {
    password: {
      value: '',
      isValid: false,
    },
    email: {
      value: '',
      isValid: false,
    },
    errorCode: '',
  }
  constructor(props) {
    super(props)
    this.handlePasswordChange = this.handlePasswordChange.bind(this)
    this.handleEmailChange = this.handleEmailChange.bind(this)
    this.handleSubmit = this.handleSubmit.bind(this)
  }
  handlePasswordChange(event) {
    const password = event.target.value
    const isValid = testPasswordStrength(password)
    this.setState(prevState => ({
      ...prevState,
      password: {
        isValid,
        value: password,
      },
    }))
  }
  handleEmailChange(event) {
    const email = event.target.value
    const isValid = EmailValidator.validate(email)
    this.setState(prevState => ({
      ...prevState,
      email: {
        isValid,
        value: email,
      },
    }))
  }
  async handleSubmit(e) {
    e.preventDefault()
    let formValid = true

    if (!this.state.email.isValid) {
      formValid = false
      this.setState(prevState => ({
        ...prevState,
        errorCode: 'email',
      }))
    }

    if (!this.state.password.isValid) {
      formValid = false
      this.setState(prevState => ({
        ...prevState,
        errorCode: 'password',
      }))
    }

    if (formValid) {
      try {
        const { data } = await axios.post('/auth/register', {
          email: this.state.email.value,
          password: this.state.password.value,
        })
        this.props.setAuthenticatedUser({
          authed: data.auth,
          user: {
            email: this.state.email.value,
            id: data.id,
          },
        })
        setAuthCookie(data.token)
      } catch (error) {
        // SENTRY will handle this
        console.error('error', error)
      }
    }
  }
  render() {
    return (
      <form onSubmit={this.handleSubmit} data-testid="register-user-form">
        <input
          type="email"
          placeholder="Email"
          name="email"
          data-testid="register-user-form-email"
          value={this.state.email.value}
          required
          onChange={this.handleEmailChange}
        />
        <input
          type="password"
          placeholder="Password"
          name="password"
          value={this.state.password.value}
          data-testid="register-user-form-password"
          required
          onChange={this.handlePasswordChange}
        />
        <button type="submit">Create Account</button>
        <FormError errorCode={this.state.errorCode} />
      </form>
    )
  }
}

function FormError({ errorCode }) {
  if (errorCode === 'email') {
    return (
      <div className="formError" data-testid="form-error-email">
        Please enter a valid email address.
      </div>
    )
  } else if (errorCode === 'password') {
    return (
      <div className="formError" data-testid="form-error-password">
        Please enter a valid password.
        <ul>
          <li>Must be at least 6 characters.</li>
          <li>Must have at least one lowercase letter.</li>
          <li>Must have at least one uppercase letter.</li>
          <li>Must have at least one number.</li>
        </ul>
      </div>
    )
  }
  return null
}

FormError.propTypes = {
  errorCode: PropTypes.string.isRequired,
}

const mapStateToProps = () => ({})

const mapDispatchToProps = dispatch =>
  bindActionCreators(authActionCreators, dispatch)

export default connect(mapStateToProps, mapDispatchToProps)(RegisterUserForm)
