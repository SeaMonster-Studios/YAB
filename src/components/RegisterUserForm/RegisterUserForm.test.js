import React from 'react'
import { Simulate, wait, render } from 'react-testing-library'
import faker from 'faker'
//
import { renderWithRedux } from '../../utils/testing/testing'
import RegisterUserForm, {
  RegisterUserForm as RegisterUserFormSansRedux,
} from './RegisterUserForm'

describe('RegisterUserForm Component Test', () => {
  it('Renders with correct default values from redux store', () => {
    const wrapper = renderSetupWithRedux()

    // make sure it renders
    expect(wrapper.container).toMatchSnapshot()

    // No user should be logged in if this form is showing
    expect(wrapper.initialState.auth.authed).toBe(false)
    expect(wrapper.initialState.auth.user.email).toBe('')
    expect(wrapper.initialState.auth.user.id).toBe('')
  })

  it('Expects valid email input value to update on user input', () => {
    const { email } = renderSetup()
    const value = faker.internet.email()
    Simulate.change(email, { target: { value } })

    expect(email.value).toBe(value)
  })

  it('Expects valid password input value to update on user input', () => {
    const { password } = renderSetup()
    const value = faker.internet.password(100)
    Simulate.change(password, { target: { value } })

    expect(password.value).toBe(value)
  })

  describe('Expects correct error message to be displayed when email input is invalid', () => {
    const invalidEmails = [
      faker.random.word(),
      faker.random.words(),
      faker.internet.url(),
      faker.internet.userName(),
      faker.internet.domainName(),
      'test@adfasd',
      'something.com',
    ]

    invalidEmails.forEach(invalidEmail => {
      it(`Expects "${invalidEmail}" to be invalid, resulting in the email error message being displayed`, async () => {
        const { email, password, getByTestId, store } = renderSetupWithRedux()
        let formError

        Simulate.change(email, { target: { value: invalidEmail } })
        Simulate.change(password, {
          target: { value: faker.internet.password(100) },
        })

        await wait(() => Simulate.submit(getByTestId('register-user-form')))

        const stateAfterSubmit = store.getState()

        await wait(() => (formError = getByTestId('form-error-email')))

        expect(formError).toBeDefined()
        expect(stateAfterSubmit.auth.authed).toBe(false)
      })
    })
  })

  describe('Expects correct error message to be displayed when password input is invalid', () => {
    const invalidPasswords = [
      faker.random.word(),
      faker.random.words(),
      faker.internet.password(5), // needs at least 6 characters
      'THISISMYPASSW0RD!32423', // needs lowercase
      'test0320', // needs uppercase
      'Password', // needs number
    ]

    invalidPasswords.forEach(invalidPassword => {
      it(`Expects "${invalidPassword}" to be invalid, resulting in the password error message being displayed`, async () => {
        const { email, password, getByTestId, store } = renderSetupWithRedux()
        let formError

        Simulate.change(email, { target: { value: faker.internet.email() } })
        Simulate.change(password, {
          target: { value: invalidPassword },
        })

        await wait(() => Simulate.submit(getByTestId('register-user-form')))

        const stateAfterSubmit = store.getState()

        await wait(() => (formError = getByTestId('form-error-password')))

        expect(formError).toBeDefined()
        expect(stateAfterSubmit.auth.authed).toBe(false)
      })
    })
  })
})

function renderSetupWithRedux(overrides) {
  const wrapper = renderWithRedux(<RegisterUserForm />)
  const email = wrapper.getByTestId('register-user-form-email')
  const password = wrapper.getByTestId('register-user-form-password')
  const initialState = wrapper.store.getState()

  return {
    ...wrapper,
    email,
    password,
    initialState,
    ...overrides,
  }
}

function renderSetup(overrides) {
  const wrapper = render(
    <RegisterUserFormSansRedux setAuthenticatedUser={() => ({})} />,
  )
  const email = wrapper.getByTestId('register-user-form-email')
  const password = wrapper.getByTestId('register-user-form-password')

  return {
    ...wrapper,
    email,
    password,
    ...overrides,
  }
}
