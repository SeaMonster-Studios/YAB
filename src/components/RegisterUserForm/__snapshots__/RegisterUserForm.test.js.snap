// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RegisterUserForm Component Test Renders with correct default values from redux store 1`] = `
<div>
  <form
    data-testid="register-user-form"
  >
    <input
      data-testid="register-user-form-email"
      name="email"
      placeholder="Email"
      required=""
      type="email"
      value=""
    />
    <input
      data-testid="register-user-form-password"
      name="password"
      placeholder="Password"
      required=""
      type="password"
      value=""
    />
    <button
      type="submit"
    >
      Create Account
    </button>
  </form>
</div>
`;
